custom_metrics/almanac.js

```diff
@@ -7,37 +7,233 @@
+
+          if (document.location.hostname === link.hostname) {
+            internal++;
+            if (document.location.pathname === link.pathname && link.hash.length > 1) {
+              // check if hash matches an element in the DOM (scroll to)
+              try {
+                var element = document.querySelector(link.hash);
+                if (element) {
+                  hash++;
+                } else {
+                  navigateHash++;
+                }
+              } catch (e) {}
             }
+          } else if (document.location.hostname !== link.hostname) {
+            external++;
+          }
+        }
+      }
+    }
+
+    return { internal, external, hash, navigateHash };
+  })(),
+  // Extracts titles used and counts the words, to flag thin content pages
+  'seo-titles': (() => {
+    //metric 10.9
+    var nodes = document.querySelectorAll('h1, h2, h3, h4');
+    var titleWords = -1;
+    var titleElements = -1;
+
+    if (nodes) {
+      titleWords = 0;
+      titleElements = 0;
 
-            linkNodes.push(el);
+      // analyse each title node
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+        // remove extra whitespace
+        var nodeText = node.textContent.replace(/\s+/g, ' ').replace(/^\s+|\s+$/, '');
+        var nodeWordsCount = nodeText.split(' ').length;
+
+        if (nodeWordsCount > 0) {
+          titleWords += nodeWordsCount;
+          titleElements++;
         }
+      }
+    }
+    return { titleWords, titleElements };
+  })(),
+  // Extracts words on the page to flag thin content pages
+  'seo-words': (() => {
+    //metric 10.9
+    function analyseTextNode(node) {
+      // remove extra whitespace
+      var nodeText = node.textContent.replace(/\s+/g, ' ').replace(/^\s+|\s+$/, '');
+      var nodeWordsCount = nodeText.split(' ').length;
+
+      if (nodeWordsCount > 3) {
+        // update counts
+        wordsCount += nodeWordsCount;
+        wordElements++;
+      }
+    }
+
+    var body = document.querySelector('body');
+    var wordsCount = -1;
+    var wordElements = -1;
+    if (body) {
+      wordsCount = 0;
+      wordElements = 0;
+      var n,
+        nodes = [],
+        walk = document.createTreeWalker(
+          body,
+          NodeFilter.SHOW_ALL,
+          {
+            acceptNode: function(node) {
+              if (node.nodeName === 'STYLE' || node.nodeName === 'SCRIPT') {
+                return NodeFilter.FILTER_REJECT;
+              }
+              if (node.nodeType !== 3) {
+                return NodeFilter.FILTER_SKIP;
+              }
+              return NodeFilter.FILTER_ACCEPT;
+            }
+          },
+          false
+        );
+      while ((n = walk.nextNode())) analyseTextNode(n);
+    }
+    return { wordsCount, wordElements };
+  })(),
+  // Parse <input> elements
+  'input-elements': (() => {
+    // Used by  12.12, 12.14
+    var nodes = document.querySelectorAll('body input, body select');
```

Comment:
I don't think the body part of the selector is necessary. input/select can't be found anywhere else (valid, at least).
```suggestion
    var nodes = document.querySelectorAll('input, select');
```